pkg/cloud/azure/actuators/machineset/controller.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	machinev1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/record"
+	providerconfigv1 "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/actuators"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+const (
+	cpuKey    = "machine.openshift.io/vCPU"
+	memoryKey = "machine.openshift.io/memoryMb"
+	gpuKey    = "machine.openshift.io/GPU"
+)
+
+type MachineSetReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+func (r *MachineSetReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&machinev1.MachineSet{}).
+		WithOptions(options).
+		Build(r)
+
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	r.recorder = mgr.GetEventRecorderFor("machineset-controller")
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+func (r *MachineSetReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	logger := r.Log.WithValues("machineset", req.Name, "namespace", req.Namespace)
+	logger.V(3).Info("Reconciling")
+
+	ctx := context.Background()
+	machineSet := &machinev1.MachineSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, machineSet); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Ignore deleted MachineSets, this can happen when foregroundDeletion
+	// is enabled
+	if !machineSet.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	originalMachineSetToPatch := client.MergeFrom(machineSet.DeepCopy())
+	defer func() {
```

Comment:
we we need to defer here? 
what's the benefit over calling this explicitly at the end of this func? I'd find that more readable. wdyt?

Reply:
I did this because I had meant to put a  return in the `if err != nil` block on lines 96-99, but I seemingly forgot to do that.
I was following a pattern from upstream but happy to rework it to avoid the defer

Reply:
Added a commit to rework this, gonna squash it before merge, WDYT?

Reply:
sgtm. Please keep revendor commit separate